src/main/java/com/es/phoneshop/web/ProductListPageServlet.java

```diff
@@ -45,8 +56,48 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
                     SortField.getSortFieldByRequestParam(sortField),
                     SortOrder.getSortOrderByRequestParam(sortOrder)));
         }
-        request.setAttribute(RECENTLY_VIEW_ATTRIBUTE, recentlyViewService.getRecentlyViewedProducts(request.getSession()));
-        request.getRequestDispatcher(JSP_PATH).forward(request, response);
+        request.setAttribute(RECENTLY_VIEW_ATTRIBUTE,
+                recentlyViewService.getRecentlyViewedProducts(request.getSession()));
+        request.getRequestDispatcher(PRODUCT_LIST_JSP_PATH).forward(request, response);
     }
 
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String code = request.getParameter(PRODUCT_CODE_PARAM);
+        String quantityString = request.getParameter(QUANTITY_PARAM);
+        String query = request.getParameter(QUERY_PARAM);
+        Cart cart = cartService.getCart(request.getSession());
+        try {
+            cartService.add(cart, code, parseQuantity(quantityString, request));
+            if (query.isEmpty()) {
+                response.sendRedirect(String.format("%s/products?message=Added to cart successfully",
+                        request.getContextPath()));
+            } else {
+                response.sendRedirect(String.format("%s/products?message=Added to cart successfully&query=%s",
+                        request.getContextPath(),
+                        query));
+            }
+        } catch (Exception e) {
+            request.setAttribute(ERROR_ATTRIBUTE, getMessageException(e));
+            doGet(request, response);
+            return;
+        }
+    }
+
+    private int parseQuantity(String quantityString, HttpServletRequest request) throws ParseException {
+        NumberFormat format = NumberFormat.getInstance(request.getLocale());
+        return format.parse(quantityString).intValue();
+    }
+
+
+    private String getMessageException(Exception e) {
+        if (e.getClass() == ParseException.class) {
```

Comment:
Why not instanceof?

Reply:
**getClass** is faster than **instanceof**. If I am not mistaken, then **getClass** does not look at the inheritance hierarchy and will be more specific.